Added-By: Gordon Malm <gengor@gentoo.org>

Note: Backported to earlier kernels.  Original message below.

---

From jejb@kernel.org  Tue Nov 11 10:15:37 2008
From: Maciej Sosnowski <maciej.sosnowski@intel.com>
Date: Tue, 11 Nov 2008 17:50:09 GMT
Subject: I/OAT: fix channel resources free for not allocated channels
To: stable@kernel.org
Message-ID: <200811111750.mABHo9IU025655@hera.kernel.org>

From: Maciej Sosnowski <maciej.sosnowski@intel.com>

commit c3d4f44f50b65b0b0290e357f8739cfb3f4bcaca upstream

If the ioatdma driver is loaded but not used it does not allocate descriptors.
Before it frees channel resources it should first be sure
that they have been previously allocated.

Signed-off-by: Maciej Sosnowski <maciej.sosnowski@intel.com>
Tested-by: Tom Picard <tom.s.picard@intel.com>
Signed-off-by: Dan Williams <dan.j.williams@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/dma/ioat_dma.c |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/drivers/dma/ioat_dma.c
+++ b/drivers/dma/ioat_dma.c
@@ -524,6 +524,12 @@ static void ioat_dma_free_chan_resources
 	struct ioat_desc_sw *desc, *_desc;
 	int in_use_descs = 0;
 
+	/* Before freeing channel resources first check
+	* if they have been previously allocated for this channel.
+	*/
+	if (ioat_chan->desccount == 0)
+		return;
+
 	tasklet_disable(&ioat_chan->cleanup_task);
 	ioat_dma_memcpy_cleanup(ioat_chan);
 
@@ -585,6 +591,7 @@ static void ioat_dma_free_chan_resources
 	ioat_chan->last_completion = ioat_chan->completion_addr = 0;
 	ioat_chan->pending = 0;
 	ioat_chan->dmacount = 0;
+	ioat_chan->desccount = 0;
 }
 
 /**
